api/v1beta1/clusterclass_types.go

```diff
@@ -553,11 +553,43 @@ type LocalObjectTemplate struct {
 
 // ClusterClassStatus defines the observed state of the ClusterClass.
 type ClusterClassStatus struct {
+	// Variables is a list of ClusterClassStatusVariable that are defined for the ClusterClass.
+	// +optional
+	Variables []ClusterClassStatusVariable `json:"variables,omitempty"`
+
 	// Conditions defines current observed state of the ClusterClass.
 	// +optional
 	Conditions Conditions `json:"conditions,omitempty"`
 }
 
+// ClusterClassStatusVariable defines a variable which appears in the status of a ClusterClass.
+type ClusterClassStatusVariable struct {
+	// Name is the name of the variable.
+	Name string `json:"name"`
+
+	// RequiresNamespace specifies whether or not a variable must be addressed by namespace when a value is provided for
+	// it in a Cluster.
+	RequiresNamespace bool `json:"requiresNamespace"`
+
+	// Definitions is a list of variables with a schema for each namespace the variable is defined in.
+	Definitions []ClusterClassStatusVariableDefinition `json:"definitions"`
+}
+
+// ClusterClassStatusVariableDefinition defines a variable which appears in the status of a ClusterClass.
+type ClusterClassStatusVariableDefinition struct {
+	// Namespace specifies the namespace of the variable definition.
+	Namespace string `json:"namespace"`
```

Comment:
The more I see this, and write it, the more I think overloading the term namespace might be confusing. In the current API we don't have a "global" namespace, just `non-Namespaced if possible` so I don't think namespace is even the best fit right now.

What about:
- `source`
- `scope`
- `origin`

Any other ideas? I'm not too picky :smile:, but namespace does seem like it might confuse more than help. I think something like `scope` might be the best fit.

Reply:
I am +1 to `scope` over source and origin. For the cluster author, the target audience for the `ClusterClassStatusVariableDefinition`, this information shows the where the variable will be used - the scope of the variable. 
It might be a subtle difference but the cluster author cares more about where/how the variable will be used (scope) rather than where the variable came from (source/origin).

Reply:
In my opinion namespace works well and is the best fit for what we want to express (which is that the variable names are namespaced). I think it's very unlikely that users will confuse .variables[].namespace with Kubernetes namespaces.

I assume we want to use the same name here as we want in the Cluster.spec.topology.

`source`  and `origin` would be confusing in Cluster.spec.topology. It feels like we're leaking through details of where we got the variables from to our "end users" who should only care about which variables exist for a Cluster and how to set them.

I would prefer to not use `scope` as that limits our options going forward to use scope to actually scope down variables in a different dimension (https://github.com/kubernetes-sigs/cluster-api/issues/6706). E.g. have variables with CP/InfraCluster/MD scope

Reply:
Alternative option is something like `fromPatch`. This might possibly make the whole thing easier for users to understand from reading and writing objects too.

Reply:
I think namespace is the right name for the field. I don't really see the problem

Reply:
Let's resolve this conversation as the new names are different now.